ledger/src/blockstore.rs

```diff
@@ -3718,17 +3771,42 @@ impl Blockstore {
                     warn!("Missing shred for {slot} index {shred_index}");
                     BlockstoreError::MissingShred(slot, shred_index)
                 })?;
-                shred::layout::get_merkle_root(&shred_bytes).ok_or_else(|| {
-                    let shred_index = start_index + u64::try_from(offset).unwrap();
-                    warn!("Found legacy shred for {slot}, index {shred_index}");
-                    BlockstoreError::LegacyShred(slot, shred_index)
-                })
+                let is_retransmitter_signed = shred::layout::is_retransmitter_signed(&shred_bytes)
+                    .map_err(|_| {
+                        let shred_index = start_index + u64::try_from(offset).unwrap();
+                        warn!("Found legacy shred for {slot}, index {shred_index}");
+                        BlockstoreError::LegacyShred(slot, shred_index)
+                    })?;
+                let merkle_root =
+                    shred::layout::get_merkle_root(&shred_bytes).ok_or_else(|| {
+                        let shred_index = start_index + u64::try_from(offset).unwrap();
+                        warn!("Unable to read merkle root for {slot}, index {shred_index}");
+                        BlockstoreError::MissingMerkleRoot(slot, shred_index)
+                    })?;
+                Ok((merkle_root, is_retransmitter_signed))
             })
-            .dedup_by(|res1, res2| res1.as_ref().ok() == res2.as_ref().ok())
-            .collect::<Result<Vec<Hash>>>()?;
+            .process_results::<_, _, BlockstoreError, _>(|mut iter| {
+                // lift and dedup merkle root and retransmitter individually to avoid allocation
+                // and still be able to report metrics accurately for each check.
+                let Some((merkle_root, retransmitter_signed)) = iter.next() else {
+                    return (None, false);
+                };
+                iter.fold(
```

Comment:
this is to avoid the intermediate allocation but if it's more readable, we can remove the `process_results` and instead do something like:
```
let (merkle_roots, retransmitter_signeds) = self.data_shred_cf.[...].collect().into_iter().unzip();
let is_full = merkle_roots.into_iter().dedup().collect().len() == 1;
let is_retransmitter_signed = retransmitter_signeds.into_iter().all(|rs| rs);
```

Reply:
I think the old code would still work.
i.e. you would still dedup_by as in the old code, and then add extra check at the end that the single unique element has true for `is_retransmitter_signed`.

Reply:
agreed but I was thinking it might be useful to have the metrics be accurate when this fails. For example after dedup if we had:
`[(<Hash1>, true), (<Hash1>, false)]` then the old code would have `is_full: false` and `is_retransmitter_signed: false` even though the merkle roots are the same.
Similarly something like:
`[(<Hash1>, true), (<Hash2>, true)]` means we are comprised of retransmitter shreds, however `is_retransmitter_signed` would be false. 

but perhaps this level of granularity isn't worth the extra code?

Reply:
I think the goal here should be to just add metrics which indicate the roll-out status and if we are ready to activate the features or not. So probably don't need to worry about distinguishing combinations which don't happen unless someone goes in and intentionally makes invalid changes to the shred generation code.

For example, this

> `[(<Hash1>, true), (<Hash1>, false)]` then the old code would have `is_full: false` and `is_retransmitter_signed: false` even though the merkle roots are the same.

Shouldn't happen unless someone changes the code to only generate data shreds, because the size of erasure encoded slice in "resigned" variants is different from non-resigned variants (obviously the buffer for retransmitter's signature cannot be erasure encoded). But you cannot feed in shards of different sizes to Reed-Solomon encoder so you cannot generate coding shreds.